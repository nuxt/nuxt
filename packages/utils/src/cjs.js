import { join } from 'path'
import _createRequire from 'create-require'
import jiti from 'jiti'

export const createRequire = (filename, useJiti = global.__NUXT_DEV__) => {
  if (useJiti && typeof jest === 'undefined') {
    return jiti(filename)
  }

  return _createRequire(filename)
}

const _require = createRequire()

export function isHMRCompatible (id) {
  return !/[/\\]mongoose[/\\/]/.test(id)
}

export function isExternalDependency (id) {
  return /[/\\]node_modules[/\\]/.test(id)
}

export function clearRequireCache (id) {
  if (isExternalDependency(id) && isHMRCompatible(id)) {
    return
  }

  const entry = getRequireCacheItem(id)

  if (!entry) {
    delete _require.cache[id]
    return
  }

  if (entry.parent) {
    entry.parent.children = entry.parent.children.filter(e => e.id !== id)
  }

  // Needs to be cleared before children, to protect against circular deps (#7966)
  delete _require.cache[id]

  for (const child of entry.children) {
    clearRequireCache(child.id)
  }
}

export function scanRequireTree (id, files = new Set()) {
  if (isExternalDependency(id) || files.has(id)) {
    return files
  }

  const entry = getRequireCacheItem(id)

  if (!entry) {
    files.add(id)
    return files
  }

  files.add(entry.id)

  for (const child of entry.children) {
    scanRequireTree(child.id, files)
  }

  return files
}

export function getRequireCacheItem (id) {
  try {
    return _require.cache[id]
  } catch (e) {
  }
}

export function resolveModule (id, paths) {
  if (typeof paths === 'string') {
    paths = [paths]
  }
  return _require.resolve(id, {
    paths: [].concat(...(global.__NUXT_PREPATHS__ || []), paths || [], global.__NUXT_PATHS__ || [], process.cwd())
  })
}

export function requireModule (id, paths) {
  return _require(resolveModule(id, paths))
}

export function tryRequire (id, paths) {
  try { return requireModule(id, paths) } catch (e) { }
}

export function tryResolve (id, paths) {
  try { return resolveModule(id, paths) } catch (e) { }
}

export function getPKG (id, paths) {
  return tryRequire(join(id, 'package.json'), paths)
}
