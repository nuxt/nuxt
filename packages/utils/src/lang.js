import { hasProtocol, joinURL } from 'ufo'

export const encodeHtml = function encodeHtml (str) {
  return str.replace(/</g, '&lt;').replace(/>/g, '&gt;')
}

export const isString = obj => typeof obj === 'string' || obj instanceof String

export const isNonEmptyString = obj => Boolean(obj && isString(obj))

export const isPureObject = obj => !Array.isArray(obj) && typeof obj === 'object'

export const isUrl = function isUrl (url) {
  return hasProtocol(url, true)
}

export const urlJoin = joinURL

/**
 * Wraps value in array if it is not already an array
 *
 * @param  {any} value
 * @return {array}
 */
export const wrapArray = value => Array.isArray(value) ? value : [value]

const WHITESPACE_REPLACEMENTS = [
  [/[ \t\f\r]+\n/g, '\n'], // strip empty indents
  [/{\n{2,}/g, '{\n'], // strip start padding from blocks
  [/\n{2,}([ \t\f\r]*})/g, '\n$1'], // strip end padding from blocks
  [/\n{3,}/g, '\n\n'], // strip multiple blank lines (1 allowed)
  [/^\n+/, ''], // strip blank lines at the beginning of a string
  [/\n{2,}$/, '\n'] // strip blank lines at the end of a string
]

export const stripWhitespace = function stripWhitespace (string) {
  WHITESPACE_REPLACEMENTS.forEach(([regex, newSubstr]) => {
    string = string.replace(regex, newSubstr)
  })
  return string
}
