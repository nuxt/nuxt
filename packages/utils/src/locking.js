import path from 'path'
import consola from 'consola'
import hash from 'hash-sum'
import fs from 'fs-extra'
import properlock from 'proper-lockfile'
// eslint-disable-next-line import/namespace
import { onExit } from 'signal-exit'

export const lockPaths = new Set()

export const defaultLockOptions = {
  stale: 30000,
  onCompromised: err => consola.warn(err)
}

export function getLockOptions (options) {
  return Object.assign({}, defaultLockOptions, options)
}

export function createLockPath ({ id = 'nuxt', dir, root }) {
  const sum = hash(`${root}-${dir}`)

  return path.resolve(root, 'node_modules/.cache/nuxt', `${id}-lock-${sum}`)
}

export async function getLockPath (config) {
  const lockPath = createLockPath(config)

  // the lock is created for the lockPath as ${lockPath}.lock
  // so the (temporary) lockPath needs to exist
  await fs.ensureDir(lockPath)

  return lockPath
}

export async function lock ({ id, dir, root, options }) {
  const lockPath = await getLockPath({ id, dir, root })

  try {
    const locked = await properlock.check(lockPath)
    if (locked) {
      consola.fatal(`A lock with id '${id}' already exists on ${dir}`)
    }
  } catch (e) {
    consola.debug(`Check for an existing lock with id '${id}' on ${dir} failed`, e)
  }

  let lockWasCompromised = false
  let release

  try {
    options = getLockOptions(options)

    const onCompromised = options.onCompromised
    options.onCompromised = (err) => {
      onCompromised(err)
      lockWasCompromised = true
    }

    release = await properlock.lock(lockPath, options)
  } catch (e) {}

  if (!release) {
    consola.warn(`Unable to get a lock with id '${id}' on ${dir} (but will continue)`)
    return false
  }

  if (!lockPaths.size) {
    // make sure to always cleanup our temporary lockPaths
    onExit(() => {
      for (const lockPath of lockPaths) {
        fs.removeSync(lockPath)
      }
    })
  }

  lockPaths.add(lockPath)

  return async function lockRelease () {
    try {
      await fs.remove(lockPath)
      lockPaths.delete(lockPath)

      // release as last so the lockPath is still removed
      // when it fails on a compromised lock
      await release()
    } catch (e) {
      if (!lockWasCompromised || !e.message.includes('already released')) {
        consola.debug(e)
        return
      }

      // proper-lockfile doesnt remove lockDir when lock is compromised
      // removing it here could cause the 'other' process to throw an error
      // as well, but in our case its much more likely the lock was
      // compromised due to mtime update timeouts
      const lockDir = `${lockPath}.lock`
      if (await fs.exists(lockDir)) {
        await fs.remove(lockDir)
      }
    }
  }
}
