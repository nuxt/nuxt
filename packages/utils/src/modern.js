import UAParser from 'ua-parser-js'
import semver from 'semver'

export const ModernBrowsers = {
  Edge: '16',
  Firefox: '60',
  Chrome: '61',
  'Chrome Headless': '61',
  Chromium: '61',
  Iron: '61',
  Safari: '10.1',
  Opera: '48',
  Yandex: '18',
  Vivaldi: '1.14',
  'Mobile Safari': '10.3'
}

let __modernBrowsers

const getModernBrowsers = () => {
  if (__modernBrowsers) {
    return __modernBrowsers
  }

  __modernBrowsers = Object.keys(ModernBrowsers)
    .reduce((allBrowsers, browser) => {
      allBrowsers[browser] = semver.coerce(ModernBrowsers[browser])
      return allBrowsers
    }, {})
  return __modernBrowsers
}

export const isModernBrowser = (ua) => {
  if (!ua) {
    return false
  }
  const { browser } = UAParser(ua)
  const browserVersion = semver.coerce(browser.version)
  if (!browserVersion) {
    return false
  }
  const modernBrowsers = getModernBrowsers()
  return Boolean(modernBrowsers[browser.name] && semver.gte(browserVersion, modernBrowsers[browser.name]))
}

export const isModernRequest = (req, modernMode = false) => {
  if (modernMode === false) {
    return false
  }

  const { socket = {}, headers } = req
  if (socket._modern === undefined) {
    const ua = headers && headers['user-agent']
    socket._modern = isModernBrowser(ua)
  }

  return socket._modern
}

// https://gist.github.com/samthor/64b114e4a4f539915a95b91ffd340acc
export const safariNoModuleFix = '!function(){var e=document,t=e.createElement("script");if(!("noModule"in t)&&"onbeforeload"in t){var n=!1;e.addEventListener("beforeload",function(e){if(e.target===t)n=!0;else if(!e.target.hasAttribute("nomodule")||!n)return;e.preventDefault()},!0),t.type="module",t.src=".",e.head.appendChild(t),t.remove()}}();'
