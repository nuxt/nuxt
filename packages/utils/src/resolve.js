import path from 'path'
import consola from 'consola'
import { escapeRegExp } from 'lodash'

export const startsWithAlias = aliasArray => str => aliasArray.some(c => str.startsWith(c))

export const startsWithSrcAlias = startsWithAlias(['@', '~'])

export const startsWithRootAlias = startsWithAlias(['@@', '~~'])

export const isWindows = process.platform.startsWith('win')

export const wp = function wp (p = '') {
  if (isWindows) {
    return p.replace(/\\/g, '\\\\')
  }
  return p
}

// Kept for backward compat (modules may use it from template context)
export const wChunk = function wChunk (p = '') {
  return p
}

const reqSep = /\//g
const sysSep = escapeRegExp(path.sep)
const normalize = string => string.replace(reqSep, sysSep)

export const r = function r (...args) {
  const lastArg = args[args.length - 1]

  if (startsWithSrcAlias(lastArg)) {
    return wp(lastArg)
  }

  return wp(path.resolve(...args.map(normalize)))
}

export const relativeTo = function relativeTo (...args) {
  const dir = args.shift()

  // Keep webpack inline loader intact
  if (args[0].includes('!')) {
    const loaders = args.shift().split('!')

    return loaders.concat(relativeTo(dir, loaders.pop(), ...args)).join('!')
  }

  // Resolve path
  const resolvedPath = r(...args)

  // Check if path is an alias
  if (startsWithSrcAlias(resolvedPath)) {
    return resolvedPath
  }

  // Make correct relative path
  let rp = path.relative(dir, resolvedPath)
  if (rp[0] !== '.') {
    rp = '.' + path.sep + rp
  }

  return wp(rp)
}

export function defineAlias (src, target, prop, opts = {}) {
  const { bind = true, warn = false } = opts

  if (Array.isArray(prop)) {
    for (const p of prop) {
      defineAlias(src, target, p, opts)
    }
    return
  }

  let targetVal = target[prop]
  if (bind && typeof targetVal === 'function') {
    targetVal = targetVal.bind(target)
  }

  let warned = false

  Object.defineProperty(src, prop, {
    get: () => {
      if (warn && !warned) {
        warned = true
        consola.warn({
          message: `'${prop}' is deprecated'`,
          // eslint-disable-next-line unicorn/error-message
          additional: new Error().stack.split('\n').splice(2).join('\n')
        })
      }
      return targetVal
    }
  })
}

const isIndex = s => /(.*)\/index\.[^/]+$/.test(s)

export function isIndexFileAndFolder (pluginFiles) {
  // Return early in case the matching file count exceeds 2 (index.js + folder)
  if (pluginFiles.length !== 2) {
    return false
  }
  return pluginFiles.some(isIndex)
}

export const getMainModule = () => {
  return (require && require.main) || (module && module.main) || module
}
