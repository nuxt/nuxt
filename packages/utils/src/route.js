import path from 'path'
import { get } from 'lodash'
import consola from 'consola'
import { normalizeURL, withTrailingSlash, withoutTrailingSlash } from 'ufo'
import { r } from './resolve'

const routeChildren = function (route) {
  const hasChildWithEmptyPath = route.children.some(child => child.path === '')
  if (hasChildWithEmptyPath) {
    return route.children
  }
  return [
    // Add default child to render parent page
    {
      ...route,
      children: undefined
    },
    ...route.children
  ]
}

export const flatRoutes = function flatRoutes (router, fileName = '', routes = []) {
  router.forEach((r) => {
    if ([':', '*'].some(c => r.path.includes(c))) {
      return
    }
    const route = `${fileName}${r.path}/`.replace(/\/+/g, '/')
    if (r.children) {
      return flatRoutes(routeChildren(r), route, routes)
    }

    // if child path is already absolute, do not make any concatenations
    if (r.path && r.path.startsWith('/')) {
      routes.push(r.path)
    } else if (route !== '/' && route[route.length - 1] === '/') {
      routes.push(route.slice(0, -1))
    } else {
      routes.push(route)
    }
  })
  return routes
}

// eslint-disable-next-line default-param-last
function cleanChildrenRoutes (routes, isChild = false, routeNameSplitter = '-', trailingSlash, parentRouteName) {
  const regExpIndex = new RegExp(`${routeNameSplitter}index$`)
  const regExpParentRouteName = new RegExp(`^${parentRouteName}${routeNameSplitter}`)
  const routesIndex = []
  routes.forEach((route) => {
    if (regExpIndex.test(route.name) || route.name === 'index') {
      const res = route.name.replace(regExpParentRouteName, '').split(routeNameSplitter)
      routesIndex.push(res)
    }
  })
  routes.forEach((route) => {
    route.path = isChild ? route.path.replace('/', '') : route.path
    if (route.path.includes('?')) {
      if (route.name.endsWith(`${routeNameSplitter}index`)) {
        route.path = route.path.replace(/\?\/?$/, trailingSlash ? '/' : '')
      }
      const names = route.name.replace(regExpParentRouteName, '').split(routeNameSplitter)
      const paths = route.path.split('/')
      if (!isChild) {
        paths.shift()
      } // clean first / for parents
      routesIndex.forEach((r) => {
        const i = r.indexOf('index')
        if (i < paths.length) {
          for (let a = 0; a <= i; a++) {
            if (a === i) {
              paths[a] = paths[a].replace('?', '')
            }
            if (a < i && names[a] !== r[a]) {
              break
            }
          }
        }
      })
      route.path = (isChild ? '' : '/') + paths.join('/')
    }
    route.name = route.name.replace(regExpIndex, '')
    if (route.children) {
      const defaultChildRoute = route.children.find(child => child.path === '/' || child.path === '')
      const routeName = route.name
      if (defaultChildRoute) {
        route.children.forEach((child) => {
          if (child.path !== defaultChildRoute.path) {
            const parts = child.path.split('/')
            parts[1] = parts[1].endsWith('?') ? parts[1].substr(0, parts[1].length - 1) : parts[1]
            child.path = parts.join('/')
          }
        })
        delete route.name
      }
      route.children = cleanChildrenRoutes(route.children, true, routeNameSplitter, trailingSlash, routeName)
    }
  })
  return routes
}

const DYNAMIC_ROUTE_REGEX = /^\/([:*])/

export const sortRoutes = function sortRoutes (routes) {
  routes.sort((a, b) => {
    if (!a.path.length) {
      return -1
    }
    if (!b.path.length) {
      return 1
    }
    // Order: /static, /index, /:dynamic
    // Match exact route before index: /login before /index/_slug
    if (a.path === '/') {
      return DYNAMIC_ROUTE_REGEX.test(b.path) ? -1 : 1
    }
    if (b.path === '/') {
      return DYNAMIC_ROUTE_REGEX.test(a.path) ? 1 : -1
    }

    let i
    let res = 0
    let y = 0
    let z = 0
    const _a = a.path.split('/')
    const _b = b.path.split('/')
    for (i = 0; i < _a.length; i++) {
      if (res !== 0) {
        break
      }
      y = _a[i] === '*' ? 2 : _a[i].includes(':') ? 1 : 0
      z = _b[i] === '*' ? 2 : _b[i].includes(':') ? 1 : 0
      res = y - z
      // If a.length >= b.length
      if (i === _b.length - 1 && res === 0) {
        // unless * found sort by level, then alphabetically
        res = _a[i] === '*'
          ? -1
          : (
            _a.length === _b.length ? a.path.localeCompare(b.path) : (_a.length - _b.length)
          )
      }
    }

    if (res === 0) {
      // unless * found sort by level, then alphabetically
      res = _a[i - 1] === '*' && _b[i]
        ? 1
        : (
          _a.length === _b.length ? a.path.localeCompare(b.path) : (_a.length - _b.length)
        )
    }
    return res
  })

  routes.forEach((route) => {
    if (route.children) {
      sortRoutes(route.children)
    }
  })

  return routes
}

export const createRoutes = function createRoutes ({
  files,
  srcDir,
  pagesDir = '',
  routeNameSplitter = '-',
  supportedExtensions = ['vue', 'js'],
  trailingSlash
}) {
  const routes = []
  files.forEach((file) => {
    const keys = file
      .replace(new RegExp(`^${pagesDir}`), '')
      .replace(new RegExp(`\\.(${supportedExtensions.join('|')})$`), '')
      .replace(/\/{2,}/g, '/')
      .split('/')
      .slice(1)
    const route = { name: '', path: '', component: r(srcDir, file) }
    let parent = routes
    keys.forEach((key, i) => {
      // remove underscore only, if its the prefix
      const sanitizedKey = key.startsWith('_') ? key.substr(1) : key

      route.name = route.name
        ? route.name + routeNameSplitter + sanitizedKey
        : sanitizedKey
      route.name += key === '_' ? 'all' : ''
      route.chunkName = file.replace(new RegExp(`\\.(${supportedExtensions.join('|')})$`), '')
      const child = parent.find(parentRoute => parentRoute.name === route.name)

      if (child) {
        child.children = child.children || []
        parent = child.children
        route.path = ''
      } else if (key === 'index' && i + 1 === keys.length) {
        route.path += i > 0 ? '' : '/'
      } else {
        route.path += '/' + normalizeURL(getRoutePathExtension(key))
        if (key.startsWith('_') && key.length > 1) {
          route.path += '?'
        }
      }
    })
    if (trailingSlash !== undefined) {
      route.pathToRegexpOptions = { ...route.pathToRegexpOptions, strict: true }
      if (trailingSlash && !route.path.endsWith('*')) {
        route.path = withTrailingSlash(route.path)
      } else {
        route.path = withoutTrailingSlash(route.path)
      }
    }

    parent.push(route)
  })

  sortRoutes(routes)
  return cleanChildrenRoutes(routes, false, routeNameSplitter, trailingSlash)
}

// Guard dir1 from dir2 which can be indiscriminately removed
export const guardDir = function guardDir (options, key1, key2) {
  const dir1 = get(options, key1, false)
  const dir2 = get(options, key2, false)

  if (
    dir1 &&
    dir2 &&
    (
      dir1 === dir2 ||
      (
        dir1.startsWith(dir2) &&
        !path.basename(dir1).startsWith(path.basename(dir2))
      )
    )
  ) {
    const errorMessage = `options.${key2} cannot be a parent of or same as ${key1}`
    consola.fatal(errorMessage)
    throw new Error(errorMessage)
  }
}

const getRoutePathExtension = (key) => {
  if (key === '_') {
    return '*'
  }

  if (key.startsWith('_')) {
    return `:${key.substr(1)}`
  }

  return key
}

export const promisifyRoute = function promisifyRoute (fn, ...args) {
  // If routes is an array
  if (Array.isArray(fn)) {
    return Promise.resolve(fn)
  }
  // If routes is a function expecting a callback
  if (fn.length === arguments.length) {
    return new Promise((resolve, reject) => {
      fn((err, routeParams) => {
        if (err) {
          reject(err)
        }
        resolve(routeParams)
      }, ...args)
    })
  }
  let promise = fn(...args)
  if (
    !promise ||
    (!(promise instanceof Promise) && typeof promise.then !== 'function')
  ) {
    promise = Promise.resolve(promise)
  }
  return promise
}
