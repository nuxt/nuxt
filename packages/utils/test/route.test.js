import { flatRoutes, createRoutes, guardDir, promisifyRoute } from '../src/route'

describe('util: route', () => {
  test('should flat route with path', () => {
    const routes = flatRoutes([
      { name: 'login', path: '/login' },
      { name: 'about', path: '/about' },
      {
        name: 'posts',
        path: '',
        children: [
          { name: 'posts-list', path: '' },
          { name: 'posts-create', path: 'post' }
        ]
      }
    ])
    expect(routes).toEqual(['/login', '/about', '/', '/post'])
  })

  test('should ignore route with * and :', () => {
    const routes = flatRoutes([
      { name: 'login', path: '/login' },
      { name: 'foo', path: '/foo/:id' },
      { name: 'bar', path: '/bar/*' }
    ])
    expect(routes).toEqual(['/login'])
  })

  test('should resolve route with /', () => {
    const routes = flatRoutes([
      {
        name: 'foo',
        path: '/',
        children: [
          { name: 'foo-bar', path: 'foo/bar' },
          { name: 'foo-baz', path: 'foo/baz' }
        ]
      }
    ])
    expect(routes).toEqual(['/', '/foo/bar', '/foo/baz'])
  })

  test('should flat absolute routes', () => {
    const routes = flatRoutes([
      {
        name: 'foo',
        path: '/foo',
        children: [
          { name: 'foo-bar', path: '/foo/bar' },
          { name: 'foo-baz', path: '/foo/baz' }
        ]
      }
    ])

    expect(routes).toEqual(['/foo', '/foo/bar', '/foo/baz'])
  })

  test('should flat absolute routes with empty path', () => {
    const routes = flatRoutes([
      {
        name: 'foo',
        path: '/foo',
        children: [
          { name: 'foo-root', path: '' },
          { name: 'foo-bar', path: '/foo/bar' },
          { name: 'foo-baz', path: '/foo/baz' }
        ]
      }
    ])

    expect(routes).toEqual(['/foo', '/foo/bar', '/foo/baz'])
  })

  describe('util: route guard', () => {
    test('should guard parent dir', () => {
      expect(() => {
        guardDir({ dir1: '/root/parent', dir2: '/root' }, 'dir1', 'dir2')
      }).toThrow()
    })

    test('should guard same dir', () => {
      expect(() => {
        guardDir({ dir1: '/root/parent', dir2: '/root/parent' }, 'dir1', 'dir2')
      }).toThrow()
    })

    test('should not guard same level dir', () => {
      expect(() => {
        guardDir({ dir1: '/root/parent-next', dir2: '/root/parent' }, 'dir1', 'dir2')
      }).not.toThrow()
    })

    test('should not guard same level dir - 2', () => {
      expect(() => {
        guardDir({ dir1: '/root/parent', dir2: '/root/parent-next' }, 'dir1', 'dir2')
      }).not.toThrow()
    })

    test('should not guard child dir', () => {
      expect(() => {
        guardDir({ dir1: '/root/parent', dir2: '/root/parent/child' }, 'dir1', 'dir2')
      }).not.toThrow()
    })
  })

  describe('util: route promisifyRoute', () => {
    test('should promisify array routes', () => {
      const array = [1]
      const promise = promisifyRoute(array)
      expect(typeof promise).toBe('object')
      return promise.then((res) => {
        expect(res).toBe(array)
      })
    })

    test('should promisify functional routes', () => {
      const array = [1, 2]
      const fn = function () {
        return array
      }
      const promise = promisifyRoute(fn)
      expect(typeof promise).toBe('object')
      return promise.then((res) => {
        expect(res).toBe(array)
      })
    })

    test('should promisify promisable functional routes', () => {
      const array = [1, 2, 3]
      const fn = function () {
        return new Promise((resolve) => {
          resolve(array)
        })
      }
      const promise = promisifyRoute(fn)
      expect(typeof promise).toBe('object')
      return promise.then((res) => {
        expect(res).toBe(array)
      })
    })

    test('should promisify promisable functional routes with arguments', () => {
      const fn = function (array) {
        return new Promise((resolve) => {
          resolve(array)
        })
      }
      const array = [1, 2, 3]
      const promise = promisifyRoute(fn, array)
      expect(typeof promise).toBe('object')
      return promise.then((res) => {
        expect(res).toBe(array)
      })
    })

    test('should promisify functional routes with error', () => {
      const fn = function (cb) {
        cb(new Error('Error here'))
      }
      const promise = promisifyRoute(fn)
      expect(typeof promise).toBe('object')
      return promise.catch((e) => {
        expect(e.message).toBe('Error here')
      })
    })

    test('should promisify functional routes with arguments and error', () => {
      const fn = function (cb, array) {
        cb(new Error('Error here: ' + array.join()))
      }
      const array = [1, 2, 3, 4]
      const promise = promisifyRoute(fn, array)
      expect(typeof promise).toBe('object')
      return promise.catch((e) => {
        expect(e.message).toBe('Error here: ' + array.join())
      })
    })

    test('should promisify functional routes with result', () => {
      const array = [1, 2, 3, 4]
      const fn = function (cb) {
        cb(null, array)
      }
      const promise = promisifyRoute(fn)
      expect(typeof promise).toBe('object')
      return promise.then((res) => {
        expect(res).toBe(array)
      })
    })

    test('should promisify functional routes with arguments and result', () => {
      const fn = function (cb, array, object) {
        cb(null, { array, object })
      }
      const array = [1, 2, 3, 4]
      const object = { a: 1 }
      const promise = promisifyRoute(fn, array, object)
      expect(typeof promise).toBe('object')
      return promise.then((res) => {
        expect(res.array).toBe(array)
        expect(res.object).toBe(object)
      })
    })
  })

  describe('util: route create', () => {
    const files = [
      'pages/index.vue',
      'pages/_param.vue',
      'pages/subpage/_param.vue',
      'pages/snake_case_route.vue',
      'pages/another_route/_id.vue',
      'pages/parent/index.vue',
      'pages/parent/child/index.vue',
      'pages/parent/child/test.vue'
    ]
    const srcDir = '/some/nuxt/app'
    const pagesDir = 'pages'

    test.posix('createRoutes should allow snake case routes in posix system', () => {
      const routesResult = createRoutes({ files, srcDir, pagesDir })
      expect(routesResult).toMatchSnapshot()
    })

    test.win('createRoutes should allow snake case routes in windows system', () => {
      const routesResult = createRoutes({ files, srcDir, pagesDir })
      expect(routesResult).toMatchSnapshot()
    })

    test.posix('createRoutes should enforce trailing slashes when configured to', () => {
      const routesResult = createRoutes({ files, srcDir, pagesDir, trailingSlash: true })
      expect(routesResult).toMatchSnapshot()
    })

    test.posix('createRoutes should remove trailing slashes when configured to', () => {
      const routesResult = createRoutes({ files, srcDir, pagesDir, trailingSlash: false })
      expect(routesResult).toMatchSnapshot()
    })
  })
})
