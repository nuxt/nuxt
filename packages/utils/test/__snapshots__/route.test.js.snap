// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`util: route util: route create createRoutes should allow snake case routes in posix system 1`] = `
Array [
  Object {
    "chunkName": "pages/parent/index",
    "component": "/some/nuxt/app/pages/parent/index.vue",
    "name": "parent",
    "path": "/parent",
  },
  Object {
    "chunkName": "pages/snake_case_route",
    "component": "/some/nuxt/app/pages/snake_case_route.vue",
    "name": "snake_case_route",
    "path": "/snake_case_route",
  },
  Object {
    "chunkName": "pages/parent/child/index",
    "component": "/some/nuxt/app/pages/parent/child/index.vue",
    "name": "parent-child",
    "path": "/parent/child",
  },
  Object {
    "chunkName": "pages/parent/child/test",
    "component": "/some/nuxt/app/pages/parent/child/test.vue",
    "name": "parent-child-test",
    "path": "/parent/child/test",
  },
  Object {
    "chunkName": "pages/another_route/_id",
    "component": "/some/nuxt/app/pages/another_route/_id.vue",
    "name": "another_route-id",
    "path": "/another_route/:id?",
  },
  Object {
    "chunkName": "pages/subpage/_param",
    "component": "/some/nuxt/app/pages/subpage/_param.vue",
    "name": "subpage-param",
    "path": "/subpage/:param?",
  },
  Object {
    "chunkName": "pages/index",
    "component": "/some/nuxt/app/pages/index.vue",
    "name": "index",
    "path": "/",
  },
  Object {
    "chunkName": "pages/_param",
    "component": "/some/nuxt/app/pages/_param.vue",
    "name": "param",
    "path": "/:param",
  },
]
`;

exports[`util: route util: route create createRoutes should allow snake case routes in windows system 1`] = `
Array [
  Object {
    "chunkName": "pages/parent/index",
    "component": "\\\\\\\\\\\\\\\\some\\\\\\\\nuxt\\\\\\\\app\\\\\\\\pages\\\\\\\\parent\\\\\\\\index.vue",
    "name": "parent",
    "path": "/parent",
  },
  Object {
    "chunkName": "pages/snake_case_route",
    "component": "\\\\\\\\\\\\\\\\some\\\\\\\\nuxt\\\\\\\\app\\\\\\\\pages\\\\\\\\snake_case_route.vue",
    "name": "snake_case_route",
    "path": "/snake_case_route",
  },
  Object {
    "chunkName": "pages/parent/child/index",
    "component": "\\\\\\\\\\\\\\\\some\\\\\\\\nuxt\\\\\\\\app\\\\\\\\pages\\\\\\\\parent\\\\\\\\child\\\\\\\\index.vue",
    "name": "parent-child",
    "path": "/parent/child",
  },
  Object {
    "chunkName": "pages/parent/child/test",
    "component": "\\\\\\\\\\\\\\\\some\\\\\\\\nuxt\\\\\\\\app\\\\\\\\pages\\\\\\\\parent\\\\\\\\child\\\\\\\\test.vue",
    "name": "parent-child-test",
    "path": "/parent/child/test",
  },
  Object {
    "chunkName": "pages/another_route/_id",
    "component": "\\\\\\\\\\\\\\\\some\\\\\\\\nuxt\\\\\\\\app\\\\\\\\pages\\\\\\\\another_route\\\\\\\\_id.vue",
    "name": "another_route-id",
    "path": "/another_route/:id?",
  },
  Object {
    "chunkName": "pages/subpage/_param",
    "component": "\\\\\\\\\\\\\\\\some\\\\\\\\nuxt\\\\\\\\app\\\\\\\\pages\\\\\\\\subpage\\\\\\\\_param.vue",
    "name": "subpage-param",
    "path": "/subpage/:param?",
  },
  Object {
    "chunkName": "pages/index",
    "component": "\\\\\\\\\\\\\\\\some\\\\\\\\nuxt\\\\\\\\app\\\\\\\\pages\\\\\\\\index.vue",
    "name": "index",
    "path": "/",
  },
  Object {
    "chunkName": "pages/_param",
    "component": "\\\\\\\\\\\\\\\\some\\\\\\\\nuxt\\\\\\\\app\\\\\\\\pages\\\\\\\\_param.vue",
    "name": "param",
    "path": "/:param",
  },
]
`;

exports[`util: route util: route create createRoutes should enforce trailing slashes when configured to 1`] = `
Array [
  Object {
    "chunkName": "pages/parent/index",
    "component": "/some/nuxt/app/pages/parent/index.vue",
    "name": "parent",
    "path": "/parent/",
    "pathToRegexpOptions": Object {
      "strict": true,
    },
  },
  Object {
    "chunkName": "pages/snake_case_route",
    "component": "/some/nuxt/app/pages/snake_case_route.vue",
    "name": "snake_case_route",
    "path": "/snake_case_route/",
    "pathToRegexpOptions": Object {
      "strict": true,
    },
  },
  Object {
    "chunkName": "pages/parent/child/index",
    "component": "/some/nuxt/app/pages/parent/child/index.vue",
    "name": "parent-child",
    "path": "/parent/child/",
    "pathToRegexpOptions": Object {
      "strict": true,
    },
  },
  Object {
    "chunkName": "pages/parent/child/test",
    "component": "/some/nuxt/app/pages/parent/child/test.vue",
    "name": "parent-child-test",
    "path": "/parent/child/test/",
    "pathToRegexpOptions": Object {
      "strict": true,
    },
  },
  Object {
    "chunkName": "pages/another_route/_id",
    "component": "/some/nuxt/app/pages/another_route/_id.vue",
    "name": "another_route-id",
    "path": "/another_route/:id?/",
    "pathToRegexpOptions": Object {
      "strict": true,
    },
  },
  Object {
    "chunkName": "pages/subpage/_param",
    "component": "/some/nuxt/app/pages/subpage/_param.vue",
    "name": "subpage-param",
    "path": "/subpage/:param?/",
    "pathToRegexpOptions": Object {
      "strict": true,
    },
  },
  Object {
    "chunkName": "pages/index",
    "component": "/some/nuxt/app/pages/index.vue",
    "name": "index",
    "path": "/",
    "pathToRegexpOptions": Object {
      "strict": true,
    },
  },
  Object {
    "chunkName": "pages/_param",
    "component": "/some/nuxt/app/pages/_param.vue",
    "name": "param",
    "path": "/:param/",
    "pathToRegexpOptions": Object {
      "strict": true,
    },
  },
]
`;

exports[`util: route util: route create createRoutes should remove trailing slashes when configured to 1`] = `
Array [
  Object {
    "chunkName": "pages/parent/index",
    "component": "/some/nuxt/app/pages/parent/index.vue",
    "name": "parent",
    "path": "/parent",
    "pathToRegexpOptions": Object {
      "strict": true,
    },
  },
  Object {
    "chunkName": "pages/snake_case_route",
    "component": "/some/nuxt/app/pages/snake_case_route.vue",
    "name": "snake_case_route",
    "path": "/snake_case_route",
    "pathToRegexpOptions": Object {
      "strict": true,
    },
  },
  Object {
    "chunkName": "pages/parent/child/index",
    "component": "/some/nuxt/app/pages/parent/child/index.vue",
    "name": "parent-child",
    "path": "/parent/child",
    "pathToRegexpOptions": Object {
      "strict": true,
    },
  },
  Object {
    "chunkName": "pages/parent/child/test",
    "component": "/some/nuxt/app/pages/parent/child/test.vue",
    "name": "parent-child-test",
    "path": "/parent/child/test",
    "pathToRegexpOptions": Object {
      "strict": true,
    },
  },
  Object {
    "chunkName": "pages/another_route/_id",
    "component": "/some/nuxt/app/pages/another_route/_id.vue",
    "name": "another_route-id",
    "path": "/another_route/:id?",
    "pathToRegexpOptions": Object {
      "strict": true,
    },
  },
  Object {
    "chunkName": "pages/subpage/_param",
    "component": "/some/nuxt/app/pages/subpage/_param.vue",
    "name": "subpage-param",
    "path": "/subpage/:param?",
    "pathToRegexpOptions": Object {
      "strict": true,
    },
  },
  Object {
    "chunkName": "pages/index",
    "component": "/some/nuxt/app/pages/index.vue",
    "name": "index",
    "path": "/",
    "pathToRegexpOptions": Object {
      "strict": true,
    },
  },
  Object {
    "chunkName": "pages/_param",
    "component": "/some/nuxt/app/pages/_param.vue",
    "name": "param",
    "path": "/:param",
    "pathToRegexpOptions": Object {
      "strict": true,
    },
  },
]
`;
